dali/pipeline/operator/eager_operator.h

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATOR_EAGER_OPERATOR_H_
+#define DALI_PIPELINE_OPERATOR_EAGER_OPERATOR_H_
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/pipeline/data/tensor_list.h"
+#include "dali/pipeline/operator/op_spec.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/util/batch_utils.h"
+#include "dali/pipeline/util/backend2workspace_map.h"
+#include "dali/pipeline/util/thread_pool.h"
+#include "dali/pipeline/workspace/workspace.h"
+
+namespace dali {
+
+template <typename Backend>
+std::shared_ptr<TensorList<Backend>> AsTensorList(
+    const std::shared_ptr<TensorList<Backend>> &in) {
+  return in;
+}
+
+template <typename Backend>
+std::shared_ptr<TensorList<Backend>> AsTensorList(
+    const std::shared_ptr<TensorVector<Backend>> &in) {
+  if (in->IsContiguous()) {
+    // Filled contiguous TensorVector, we can return TensorList directly.
+    return in->AsTensorList(false);
+  }
+
+  auto tl = std::make_shared<TensorList<Backend>>();
+  tl->Copy(*in);
+  return tl;
+}
+
+/**
+ * @brief Direct operator providing eager execution of an operator in Run.
+ */
+template <typename Backend>
+class DLL_PUBLIC EagerOperator {
+ public:
+  DLL_PUBLIC inline EagerOperator(const OpSpec &spec)
+      : batch_size_(spec.GetArgument<int>("max_batch_size")),
+        num_outputs_(spec.GetSchema().NumOutput()),
```

Comment:
As always is the case with DALI some ops have: CalculateOutputs and CalculateAdditionalOutputs to calculate the number of outputs. I think some of it is already used in Python layer in dali.ops, so maybe the spec rather than schema have a correct number of outputs.

Reply:
done